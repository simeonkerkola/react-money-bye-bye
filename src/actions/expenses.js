import uuid from 'uuid'
import database from '../firebase/firebase'

// ADD_EXPENSE
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense,
})

export const startAddExpense = (expenseData = {}) =>
  // Returning a function on Redux wouldn't work on default, need redux-thunk
  (dispatch) => {
    const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData // Destructure from expenseData
    const expense = { description, note, amount, createdAt }

    return database
      .ref('expenses')
      .push(expense)
      .then((ref) => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense,
          }),
        )
      })
  }

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id,
})

export const startRemoveExpense = ({ id }) => dispatch =>
  database
    .ref(`expenses/${id}`)
    .remove()
    .then(() => {
      dispatch(removeExpense({ id }))
    })
    .catch(err => console.log(err.message))

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates,
})

// SET_EXPENSES
export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses,
})

export const startSetExpenses = () => dispatch =>
  database
    .ref('expenses')
    .once('value')
    .then((snapshot) => {
      const expenses = []
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key, // value of automatically generated id
          ...childSnapshot.val(), // push everything else as well
        })
      })
      dispatch(setExpenses(expenses))
    })
