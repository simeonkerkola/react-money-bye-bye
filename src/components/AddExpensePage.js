import React from 'react'
import ExpenseForm from './ExpenseForm.js'

const AddExpensePage = () => (
  <div>
    My add expense component
    <h1>Add Expense</h1>
    <ExpenseForm />
  </div>
)
export default AddExpensePage
