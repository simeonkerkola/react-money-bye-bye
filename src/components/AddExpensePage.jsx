import React from 'react'

const AddExpensePage = () => (
  <div>My add expense component</div>
)

export default AddExpensePage
