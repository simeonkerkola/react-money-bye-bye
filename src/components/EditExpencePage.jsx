import React from 'react'

const EditExpencePage = (props) =>  {
  console.log(props);
  return (
    <div>Edit Expense with the id of {props.match.params.id}</div>
  )
}
export default EditExpencePage
