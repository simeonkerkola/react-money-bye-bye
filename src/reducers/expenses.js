// Expenses reducer
const expensesReducerDefaultState = []

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        // Unlike .push(), spread operator or .concat() doesn't change the state
        ...state,
        action.expense,
      ]
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id)
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense, // Grab all existing properties
            ...action.updates, // Override allproperties we passed down
          }
        }
        return expense
      })
    case 'SET_EXPENSES':
      return action.expenses
    default:
      return state
  }
}
