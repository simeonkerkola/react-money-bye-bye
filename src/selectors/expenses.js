import moment from 'moment'

// Get visible expenses
export default (expenses, { text, sortBy, startDate, endDate }) =>
  expenses
    .filter((expense) => {
      const startDateMatch = startDate
        ? moment(expense.createdAt).isSameOrAfter(startDate, 'day')
        : true
      const endDateMatch = endDate
        ? moment(expense.createdAt).isSameOrBefore(endDate, 'day')
        : true
      const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

      // If any of these is false, return is false and
      return startDateMatch && endDateMatch && textMatch
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1
      } else if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1
      }
      return 0
    })
