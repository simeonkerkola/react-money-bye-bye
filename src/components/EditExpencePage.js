import React from 'react'
import { connect } from 'react-redux'
import { editExpense, removeExpense } from '../actions/expenses'
import ExpenseForm from './ExpenseForm'

const EditExpencePage = props => (
  <div>
    <ExpenseForm
      expense={props.expense}
      onSubmit={(expense) => {
        props.dispatch(editExpense(props.expense.id, expense))
        props.history.push('/') // throw user to dashboard page
      }}
    />
    <button
      onClick={() => {
        props.dispatch(removeExpense({ id: props.expense.id }))
        props.history.push('/')
        console.log('click')
      }}
    >
      Remove
    </button>
  </div>
)

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id),
})

export default connect(mapStateToProps)(EditExpencePage)
