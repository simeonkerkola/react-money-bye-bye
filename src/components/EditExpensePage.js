import React from 'react'
import { connect } from 'react-redux'
import Modal from 'react-modal'
import { startEditExpense, startRemoveExpense } from '../actions/expenses'
import ExpenseForm from './ExpenseForm'
import RemoveExpenseModal from './RemoveExpenseModal'

export class EditExpensePage extends React.Component {
  state = {
    readyToBeRemoved: undefined,
  }
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense)
    this.props.history.push('/') // throw user to dashboard page
  }

  onRemove = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id })
    this.props.history.push('/')
  }

  onCloseModal = () => {
    this.setState(() => ({ readyToBeRemoved: undefined }))
  }

  askToRemove = (expense) => {
    this.setState(() => ({ readyToBeRemoved: true }))
  }
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2 className="page-header__title">Edit Expense</h2>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
          <button className="btn btn--secondary" onClick={this.askToRemove}>
            Remove
          </button>
        </div>
        <RemoveExpenseModal
          readyToBeRemoved={this.state.readyToBeRemoved}
          closeModal={this.onCloseModal}
          expense={this.props.expense}
          onRemove={this.onRemove}
        />
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id),
})
const mapDispatchToProps = dispatch => ({
  startEditExpense: (id, changes) => dispatch(startEditExpense(id, changes)),
  startRemoveExpense: id => dispatch(startRemoveExpense(id)),
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
