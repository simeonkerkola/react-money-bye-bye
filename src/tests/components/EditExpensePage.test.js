import React from 'react'
import { shallow } from 'enzyme'
import expenses from '../fixtures/expenses'
import { EditExpensePage } from '../../components/EditExpensePage'

let editExpense
let startRemoveExpense
let history
let wrapper

beforeEach(() => {
  editExpense = jest.fn()
  startRemoveExpense = jest.fn()
  history = { push: jest.fn() } // history.push('/')
  wrapper = shallow(
    <EditExpensePage
      editExpense={editExpense}
      startRemoveExpense={startRemoveExpense}
      history={history}
      expense={expenses[0]}
    />,
  )
})

test('should render EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle editExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0])
  expect(history.push).toHaveBeenCalledWith('/')
  expect(editExpense).toHaveBeenCalledWith(expenses[0].id, expenses[0])
})

test('should handle startRemoveExpense', () => {
  wrapper.find('button').simulate('click')
  expect(history.push).toHaveBeenCalledWith('/')
  expect(startRemoveExpense).toHaveBeenCalledWith({
    id: expenses[0].id,
  })
})
