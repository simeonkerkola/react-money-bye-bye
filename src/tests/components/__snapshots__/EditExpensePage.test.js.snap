// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h2
        className="page-header__title"
      >
        Edit Expense
      </h2>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 1.95,
          "createdAt": 0,
          "description": "Gum",
          "id": "1",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="btn btn--secondary"
      onClick={[Function]}
    >
      Remove
    </button>
  </div>
  <RemoveExpenseModal
    closeModal={[Function]}
    expense={
      Object {
        "amount": 1.95,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "",
      }
    }
    onRemove={[Function]}
  />
</div>
`;
